runtime/src/topo/hwloc/topo-hwloc.c

```diff
@@ -887,14 +896,219 @@ int chpl_topo_bindCPU(int id) {
+  int *assignedNics = NULL;
+
+  if (root->type != HWLOC_OBJ_PACKAGE) {
+    // We aren't running in a socket, so we don't care which NIC is used.
+    goto done;
+  }
+
+  // find the PCI object corresponding to the specified NIC
+  for (hwloc_obj_t obj = hwloc_get_next_pcidev(topology, NULL);
+       obj != NULL;
+       obj = hwloc_get_next_pcidev(topology, obj)) {
+    if (obj->type == HWLOC_OBJ_PCI_DEVICE) {
+      struct hwloc_pcidev_attr_s *attr = &(obj->attr->pcidev);
+      if ((attr->domain == inAddr->domain) && (attr->bus == inAddr->bus) &&
+          (attr->dev == inAddr->device) && (attr->func == inAddr->function)) {
+        nic = obj;
+        break;
+      }
+    }
+  }
+  if (nic == NULL) {
+    _DBG_P("Could not find NIC %04x:%02x:%02x.%x", inAddr->domain,
+           inAddr->bus, inAddr->device, inAddr->function);
+    goto done;
+  }
+
+  // Find all the NICS of the same vendor and device as the specified NIC and
+  // sort them by socket and PCI address.
+
+  nicAttr = &(nic->attr->pcidev);
+  int maxNics = hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_PCI_DEVICE);
+  CHK_ERR(nics = sys_calloc(maxNics, sizeof(*nics)));
+  int numNics = 0;
+
+  for (hwloc_obj_t obj = hwloc_get_next_pcidev(topology, NULL);
+       obj != NULL;
+       obj = hwloc_get_next_pcidev(topology, obj)) {
+
+    if (obj->type == HWLOC_OBJ_PCI_DEVICE) {
+      struct hwloc_pcidev_attr_s *attr = &(obj->attr->pcidev);
+      if ((attr->vendor_id == nicAttr->vendor_id) &&
+          (attr->device_id == nicAttr->device_id)) {
+        hwloc_obj_t sobj = hwloc_get_ancestor_obj_by_type(topology,
+                                                          HWLOC_OBJ_PACKAGE,
+                                                          obj);
+        if (sobj == NULL) {
+          _DBG_P("Could not find socket for NIC %04x:%02x:%02x.%x",
+                 attr->domain, attr->bus, attr->dev, attr->func);
+          goto done;
+        }
+        nics[numNics].socket = sobj->logical_index;
+        nics[numNics].obj = obj;
+        nics[numNics].assigned = false;
+        numNics++;
+      }
+    }
+  }
+  qsort(nics, numNics, sizeof(*nics), compareNics);
+
+  // Use the first NIC in our socket if there is one.
+
+  for (int i = 0; i < numNics; i++) {
+    if (nics[i].socket == root->logical_index) {
+      nic = nics[i].obj;
+      goto done;
+    }
+  }
+
+  // There isn't a NIC in our socket. Use the nth unassigned NIC, where
+  // n is our rank among the locales that don't have NICs, modulo
+  // the number of unassigned NICs. Otherwise use the nth assigned NIC.
+
+  int numLocales = chpl_get_num_locales_on_node();
+  CHK_ERR(assignedNics = sys_calloc(numLocales, sizeof(*assignedNics)));
+
+  for (int i = 0; i < numLocales; i++) {
+    assignedNics[i] = -1;
+  }
+
+  // Look for extra (unassigned) NICs
+
+  int numAssignedNics = 0;
+  for (int lid = 0; lid < numLocales; lid++) {
+    for (int nid = 0; nid < numNics; nid++) {
+      if (nics[nid].socket == lid) {
```

Comment:
This relies on `cpuInfoInit` densely mapping local_rank to sockets. Should we make that more explicit?

Reply:
Done. Added `getSocketNumber` that returns the socket number based on the local rank, and added a comment to `cpuInfoInit`.